lib/src/main/kotlin/com/lemonappdev/konsist/core/provider/KoFullyQualifiedNameProviderCore.kt

```diff
@@ -1,5 +1,42 @@
 package com.lemonappdev.konsist.core.provider
 
+import com.lemonappdev.konsist.api.declaration.KoAnnotationDeclaration
 import com.lemonappdev.konsist.api.provider.KoFullyQualifiedNameProvider
+import com.lemonappdev.konsist.api.provider.KoNullableTypeProvider
 
-internal interface KoFullyQualifiedNameProviderCore : KoFullyQualifiedNameProvider, KoBaseProviderCore
+internal interface KoFullyQualifiedNameProviderCore :
+    KoFullyQualifiedNameProvider,
+    KoBaseProviderCore,
+    KoContainingFileProviderCore,
+    KoNameProviderCore {
+    val textUsedToFqn: String
+        get() = name
+
+    override val fullyQualifiedName: String
+        get() {
+            val isInImports = containingFile
+                .imports
+                .map { it.name }
+                .firstOrNull { it.contains(textUsedToFqn) }
+
+            val isInFile = containingFile
+                .declarations()
+                .filterNot {
+                    if (this is KoAnnotationDeclaration) {
+                        it is KoAnnotationDeclaration
+                    } else {
+                        false
+                    }
+                }
+                .mapNotNull { (it as? KoFullyQualifiedNameProvider)?.fullyQualifiedName }
+                .firstOrNull { it.contains(textUsedToFqn) }
+
+            val additionalTextIfIsNullable = if (this is KoNullableTypeProvider && isNullable) "?" else ""
```

Comment:
We don't add `?` to a fully qualified name. `?` is part of type nullity, not the type itself

```kotlin
val className = String::class.qualifiedName  // ok
val className = String?::class.qualifiedName // error
```